reconcile/acs_rbac.py

```diff
@@ -0,0 +1,568 @@
+    key: str
+    value: str
+
+
+RoleAssignments = dict[str, list[AssignmentPair]]
+
+
+class AcsAccessScope(BaseModel):
+    name: str
+    description: str
+    clusters: list[str]
+    namespaces: list[dict[str, str]]
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, AcsAccessScope):
+            return (
+                self.name == other.name
+                and self.description == other.description
+                and self.clusters == other.clusters
+                and self.namespaces == other.namespaces
+            )
+        return False
+
+
+DEFAULT_ADMIN_SCOPE_NAME = "Unrestricted"
+DEFAULT_ADMIN_SCOPE_DESC = "Access to all clusters and namespaces"
+# map enum values defined in oidc-permission schema to system default ACS values
+PERMISSION_SET_NAMES = {
+    "admin": "Admin",
+    "analyst": "Analyst",
+    "vuln-admin": "Vulnerability Management Admin",
+}
+
+
+class AcsRole(BaseModel):
+    """
+    AcsRole is derived from oidc-permission-v1
+    Due to A-I role-1 ability to reference multiple oidc-permissions
+    the A-I role names cannot be used for the roles defined within ACS
+    """
+
+    name: str
+    description: str
+    assignments: list[AssignmentPair]
+    permission_set_name: str
+    access_scope: AcsAccessScope
+    system_default: Optional[bool]
+
+
+class AcsRbacIntegrationParams(PydanticRunParams):
+    thread_pool_size: int
+
+
+class AcsRbacIntegration(QontractReconcileIntegration[AcsRbacIntegrationParams]):
+    def __init__(self, params: AcsRbacIntegrationParams) -> None:
+        super().__init__(params)
+        self.qontract_integration = "acs_rbac"
+        self.qontract_integration_version = make_semver(0, 1, 0)
+        self.qontract_tf_prefix = "qracsrbac"
+
+    @property
+    def name(self) -> str:
+        return self.qontract_integration.replace("_", "-")
+
+    def get_acs_instance(self, query_func: Callable) -> AcsInstanceV1:
+        """
+        Get an ACS instance
+
+        :param query_func: function which queries GQL Server
+        """
+        if instances := acs_instances_query(query_func=query_func).instances:
+            # mirroring logic for gitlab instances
+            # current assumption is for appsre to only utilize one instance
+            if len(instances) != 1:
+                raise AppInterfaceSettingsError("More than one ACS instance found!")
+            return instances[0]
+        raise AppInterfaceSettingsError("No ACS instance found!")
+
+    def get_desired_state(self, query_func: Callable) -> list[AcsRole]:
+        """
+        Get desired ACS roles and associated users from App Interface
+
+        :param query_func: function which queries GQL server
+        :return: list of AcsRole derived from oidc-permission-1 definitions
+        """
+
+        query_results = acs_rbac_query(query_func=query_func).acs_rbacs
+        if query_results is None:
+            return []
+
+        desired_roles: dict[str, AcsRole] = {}
```

Comment:
Great point. I modeled this slightly different from original suggestion to avoid duplicating the permission object. This made the `build` code more intuitive in my opinion because it now only receives an object containing one `permission` and a list of users associated with the permission.

Reply:
If `permission` uniq is ensured, then ideally it's a frozen dict that hash is auto generated, can be used like `data[permission].append(user)`, but our auto generated model doesn't have that flag set, so we can inherit it to make it work:

```python

class Permission(OidcPermissionAcsV1):
    def __hash__(self):
        return hash(self.name)

...

        permission_with_users = defaultdict(list)
        for user in query_results:
            for role in user.roles or []:
                for permission in role.oidc_permissions or []:
                    if isinstance(permission, OidcPermissionAcsV1):
                        permission_with_usernames[
                            Permission(**permission.dict(by_alias=True))
                        ].append(user.org_username)
        return [
            AcsRole.build(permission, usernames)
            for permission, usernames in permission_with_usernames.items()
        ]
```

btw, you can directly use `-> Self` instead of `-> "AcsRole"`.

Reply:
Very nice! Thanks